### PR TITLE
Fix review url redirect

### PR DESCRIPTION
**Pull request by wagnerand**
Fix mozilla/addons#8453.

### CODE REVIEW

**Comment by diox** [comment]
The `FIXME` comment needs to be updated. It should probably just say something like:
```python
            # FIXME: this replace() is fragile, if the add-on slug appears
            # elsewhere in the URL it will break. Instead, we should probably
            # use `request.resolver_match` to rebuild the URL, replacing
            # `kwargs['addon_id']` if present by `addon.pk`
```

File: src/olympia/reviewers/tests/test_decorators.py [hunk 1]
```diff
@@ -34,16 +32,18 @@ def test_301_by_slug(self):
...
 
     def test_slug_replace_no_conflict(self):
-        slug = quote(self.addon.slug.encode('utf8'))
+        slug = self.addon.slug
         self.request.path = f'http://testserver/addon/{slug}/reviews/{slug}/path'
 
         res = self.view(self.request, self.addon.slug)
-        redirection = f'http://testserver/addon/{self.addon.id}/reviews/{slug}/path'
+        redirection = (
+            f'http://testserver/addon/{self.addon.id}/reviews/{quote(slug)}/path'
```
Comment (eviljeff) [inline src/olympia/reviewers/tests/test_decorators.py:45]: why is it quoted here?

File: src/olympia/reviewers/tests/test_decorators.py [hunk 2]
```diff
@@ -16,9 +16,7 @@ def setUp(self):
         self.func.__name__ = 'mock_function'
         self.view = dec.reviewer_addon_view(self.func)
         self.request = mock.Mock()
-        self.slug_path = 'http://testserver/addon/%s/reviews' % quote(
-            self.addon.slug.encode('utf-8')
-        )
+        self.slug_path = f'http://testserver/addon/{self.addon.slug}/reviews'
```
[the remark this answers is not in the document]
Comment (eviljeff) [reply src/olympia/reviewers/tests/test_decorators.py:19]: I guess not related to the changes in this PR, but made it hard to follow.

File: src/olympia/reviewers/tests/test_decorators.py [hunk 3]
```diff
@@ -34,16 +32,18 @@ def test_301_by_slug(self):
         self.assert3xx(res, self.id_path, 301)
 
     def test_301_by_guid(self):
-        self.request.path = f'http://testserver/addon/{quote(self.addon.guid)}/reviews'
+        self.request.path = f'http://testserver/addon/{self.addon.guid}/reviews'
         res = self.view(self.request, self.addon.guid)
         self.assert3xx(res, self.id_path, 301)
 
     def test_slug_replace_no_conflict(self):
```
Comment (eviljeff) [inline src/olympia/reviewers/tests/test_decorators.py:39]: what's this tests supposed to be checking for?  the slug doesn't seem to be replaced, and no conflict with what?

**Review by eviljeff — APPROVED** [review]
r+ ... because we know the fix will work and is needed.  But I have opinions on the tests.

File: src/olympia/reviewers/tests/test_decorators.py (hunk 1 above)
Comment (diox) [reply src/olympia/reviewers/tests/test_decorators.py:45]: We receive the unquoted path but we quote it (via `urlencode()`) when returning the new URL (the redirection)

File: src/olympia/reviewers/tests/test_decorators.py (hunk 3 above)
Comment (diox) [reply src/olympia/reviewers/tests/test_decorators.py:39]: If the slug is somehow present twice in the URL, we don't replace twice - said differently, we only replace the part of the URL that matters to look up the add-on.

As I've noted in the `FIXME` in the decorator (and in my comment above that rewords that `FIXME`) the whole replace  method, copied from the original `addons` decorator, is still not great and can lead to incorrect URLs depending on the slug, though.

**Comment by willdurand** [comment]
I updated this PR a bit so that we can land it and cherry-pick it. Hopefully the minor changes I made are enough.
